File: qqbot-domain/src/main/java/com/mieai/qqbot/domain/bot/QqAppId.java
package com.mieai.qqbot.domain.bot;

import java.util.Objects;

/** Identifier assigned to a bot application by QQ. */
public record QqAppId(String value) {
    public QqAppId {
        Objects.requireNonNull(value, "value must not be null");
        if (value.isBlank()) {
            throw new IllegalArgumentException("value must not be blank");
        }
        if (!value.equals(value.strip())) {
            throw new IllegalArgumentException("value must not have surrounding whitespace");
        }
        if (value.codePoints().anyMatch(Character::isWhitespace)) {
            throw new IllegalArgumentException("value must not contain whitespace");
        }
        if (value.codePoints().anyMatch(Character::isISOControl)) {
            throw new IllegalArgumentException("value must not contain control characters");
        }
    }

    public static QqAppId of(String value) {
        return new QqAppId(value);
    }

    @Override
    public String toString() {
        return value;
    }
}

File: qqbot-domain/src/main/kotlin/com/mieai/qqbot/domain/bot/QqAppId.kt
package com.mieai.qqbot.domain.bot

/** Identifier assigned to a bot application by QQ. */
data class QqAppId(val value: String) {
    init {
        require(value.isNotBlank()) { "value must not be blank" }
        require(value == value.trim()) { "value must not have surrounding whitespace" }
        require(value.codePoints().noneMatch { Character.isWhitespace(it) }) {
            "value must not contain whitespace"
        }
        require(value.codePoints().noneMatch { Character.isISOControl(it) }) {
            "value must not contain control characters"
        }
    }

    override fun toString(): String = value

    companion object {
        fun of(value: String): QqAppId = QqAppId(value)
    }
}
